mvpoly/src/monomials.rs

```diff
@@ -0,0 +1,480 @@
+            .collect();
+        // Handle the case where the result is zero because we want a unique
+        // representation
+        if monomials.is_empty() {
+            Self::zero()
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Sub<Sparse<F, N, D>> for &Sparse<F, N, D> {
+    type Output = Sparse<F, N, D>;
+
+    fn sub(self, other: Sparse<F, N, D>) -> Self::Output {
+        let mut monomials = self.monomials.clone();
+        for (exponents, coeff) in other.monomials {
+            monomials
+                .entry(exponents)
+                .and_modify(|c| *c -= coeff)
+                .or_insert(coeff);
+        }
+        // Remove monomials with zero coefficients
+        let monomials: HashMap<[usize; N], F> = monomials
+            .into_iter()
+            .filter(|(_, coeff)| !coeff.is_zero())
+            .collect();
+        // Handle the case where the result is zero because we want a unique
+        // representation
+        if monomials.is_empty() {
+            Sparse::<F, N, D>::zero()
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Sub<&Sparse<F, N, D>> for &Sparse<F, N, D> {
+    type Output = Sparse<F, N, D>;
+
+    fn sub(self, other: &Sparse<F, N, D>) -> Self::Output {
+        let mut monomials = self.monomials.clone();
+        for (exponents, coeff) in &other.monomials {
+            monomials
+                .entry(*exponents)
+                .and_modify(|c| *c -= *coeff)
+                .or_insert(*coeff);
+        }
+        // Remove monomials with zero coefficients
+        let monomials: HashMap<[usize; N], F> = monomials
+            .into_iter()
+            .filter(|(_, coeff)| !coeff.is_zero())
+            .collect();
+        // Handle the case where the result is zero because we want a unique
+        // representation
+        if monomials.is_empty() {
+            Sparse::<F, N, D>::zero()
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+/// Equality is defined as equality of the monomials.
+impl<const N: usize, const D: usize, F: PrimeField> PartialEq for Sparse<F, N, D> {
+    fn eq(&self, other: &Self) -> bool {
+        self.monomials == other.monomials
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Eq for Sparse<F, N, D> {}
+
+impl<const N: usize, const D: usize, F: PrimeField> One for Sparse<F, N, D> {
+    fn one() -> Self {
+        let mut monomials = HashMap::new();
+        monomials.insert([0; N], F::one());
+        Self { monomials }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Zero for Sparse<F, N, D> {
+    fn is_zero(&self) -> bool {
+        self.monomials.len() == 1
+            && self.monomials.contains_key(&[0; N])
+            && self.monomials[&[0; N]].is_zero()
+    }
+
+    fn zero() -> Self {
```

Comment:
why is zero not the empty map ?

Reply:
I wanted to have a single representation for the zero polynomial, and I picked the constant polynomial with `0`. No strong opinion. Feel free to change it later.